client/iotcore/client.go

```diff
@@ -0,0 +1,315 @@
+		CommonName:         commonName,
+	}
+
+	template := x509.CertificateRequest{Subject: subject, DNSNames: flag.Args()}
+	csr, err := x509.CreateCertificateRequest(rand.Reader, &template, key)
+	if err != nil {
+		return nil, err
+	}
+	csrPemBlock := &pem.Block{
+		Type:  "CERTIFICATE REQUEST",
+		Bytes: csr,
+	}
+
+	return pem.EncodeToMemory(csrPemBlock), nil
+}
+
+func (c *client) UpsertDevice(ctx context.Context,
+	sess *session.Session,
+	deviceID string,
+	device *Device,
+	policy string,
+) (*Device, error) {
+	svc := iot.New(sess)
+
+	awsDevice, err := c.GetDevice(ctx, sess, deviceID)
+	if err == nil && awsDevice != nil {
+		cert, err := svc.DescribeCertificateWithContext(ctx, &iot.DescribeCertificateInput{
+			CertificateId: aws.String(awsDevice.CertificateID),
+		})
+		if err == nil {
+			newStatus := iot.CertificateStatusInactive
+			awsDevice.Status = StatusDisabled
+			if device.Status == StatusEnabled {
+				newStatus = iot.CertificateStatusActive
+				awsDevice.Status = StatusEnabled
+			}
+
+			if *cert.CertificateDescription.Status != newStatus {
+				paramsUpdateCertificate := &iot.UpdateCertificateInput{
+					CertificateId: aws.String(awsDevice.CertificateID),
+					NewStatus:     aws.String(newStatus),
+				}
+				_, err = svc.UpdateCertificateWithContext(ctx, paramsUpdateCertificate)
+			}
+		}
+
+		return awsDevice, err
+	} else if err == ErrDeviceNotFound {
+		err = nil
+	}
+
+	var resp *iot.CreateThingOutput
+	if err == nil {
+		resp, err = svc.CreateThingWithContext(ctx,
+			&iot.CreateThingInput{
+				ThingName: aws.String(deviceID),
+			})
+	}
+
+	var respPolicy *iot.CreatePolicyOutput
+	if err == nil {
+		respPolicy, err = svc.CreatePolicyWithContext(ctx,
+			&iot.CreatePolicyInput{
+				PolicyDocument: aws.String(policy),
+				PolicyName:     aws.String(policyName(deviceID)),
+			})
+	}
+	var privKey *ecdsa.PrivateKey
+	if err == nil {
+		privKey, err = newPrivateKey()
+	}
+
+	var csr []byte
+	if err == nil {
+		csr, err = newCSR(deviceID, privKey)
+	}
+
+	var respCert *iot.CreateCertificateFromCsrOutput
+	if err == nil {
+		respCert, err = svc.CreateCertificateFromCsrWithContext(ctx,
+			&iot.CreateCertificateFromCsrInput{
+				CertificateSigningRequest: aws.String(string(csr)),
+				SetAsActive:               aws.Bool(device.Status == StatusEnabled),
+			})
+	}
+
+	if err == nil {
+		_, err = svc.AttachPolicyWithContext(ctx,
+			&iot.AttachPolicyInput{
+				PolicyName: respPolicy.PolicyName,
+				Target:     respCert.CertificateArn,
+			})
+	}
+
+	if err == nil {
+		_, err = svc.AttachThingPrincipalWithContext(ctx,
+			&iot.AttachThingPrincipalInput{
+				Principal: respCert.CertificateArn,
+				ThingName: aws.String(deviceID),
+			})
+	}
+
+	var deviceResp *Device
+	if err == nil {
+		deviceResp = &Device{
+			ID:          *resp.ThingId,
+			Name:        *resp.ThingName,
+			Status:      device.Status,
+			PrivateKey:  string(privateKeyToPem(privKey)),
+			Certificate: *respCert.CertificatePem,
+		}
```

Comment:
similar block is repeated, maybe a cool method in the Device to "load"/"create" it from resp?
why we are not passing version here?

Reply:
Honestly, I don't think so. Not too much is shared between the two code blocks, only ID/Name.